### PR TITLE
[bugfix] Check #TypeArgs == #TypeParams for TypeDefs as well as OpDefs

### PR DESCRIPTION
**Pull request by acl-cqc**
The code was in there, but seemingly in the wrong place. Not sure if my comment on `TypeArgError::WrongNumber` had mislead so I've corrected that ;)

### CODE REVIEW

File: src/resource.rs [hunk 1]
```diff
@@ -688,3 +688,48 @@ impl FromIterator<ResourceId> for ResourceSet {
...
+            ))])
+            .unwrap()
+            .into();
+        assert_eq!(typ2.tag(), TypeTag::Hashable);
+
+        // And some bad arguments...firstly, wrong kind of TypeArg:
+        def.instantiate_concrete([TypeArg::HashableType(HashableType::String)])
+            .unwrap_err();
+        // Too few arguments:
+        def.instantiate_concrete([]).unwrap_err();
```
Comment (acl-cqc) [inline src/resource.rs:727]: This one was previously panicking with an index-out-of-bounds in computing the tag

File: src/resource.rs [hunk 2]
```diff
@@ -688,3 +688,48 @@ impl FromIterator<ResourceId> for ResourceSet {
...
+        // Too many arguments:
+        def.instantiate_concrete([
+            TypeArg::ClassicType(ClassicType::F64),
+            TypeArg::ClassicType(ClassicType::F64),
+        ])
+        .unwrap_err();
```
Comment (ss2165) [inline src/resource.rs:733]: can we assert_matches on some known errors?

File: src/resource.rs (hunk 1 above)
Comment (acl-cqc) [reply src/resource.rs:727]: (If I'd provided a TypeDefTag::Explicit it would have just returned a CustomType)

File: src/resource.rs (hunk 2 above)
Comment (acl-cqc) [reply src/resource.rs:733]: We can. They'll change in an upcoming PR but yeah, ok

Comment (ss2165) [reply src/resource.rs:733]: ah, in which case some TODOs should be enough

Comment (acl-cqc) [reply src/resource.rs]: They'll be simple updates, and it'll make the change obvious and explicit, so done.

**Review by ss2165 — APPROVED** [review]
👍  thanks

**Pull request merged**